caddyhttp/basicauth/basicauth.go

```diff
@@ -63,7 +63,7 @@ func (a BasicAuth) ServeHTTP(w http.ResponseWriter, r *http.Request) (int, error
 
 			// let upstream middleware (e.g. fastcgi and cgi) know about authenticated
 			// user; this replaces the request with a wrapped instance
-			r = r.WithContext(context.WithValue(r.Context(),
+			*r = *r.WithContext(context.WithValue(r.Context(),
```

Comment:
> If I don't dereference r, then the context value is not available to the replacer.

One clue is in the standard library source for [WithContext](https://golang.org/src/net/http/request.go?s=11679:11738#L317). The `context.WithValue()` call returns a wrapped context containing all values and cancellation functions plus the new specified value. Then the call to `r.WithContext()` returns a shallow copy of the request receiver with the new, recently extended, context. The pointer dereference occurs in `r.WithContext()`. I believe the original line, `r = r.WithContext()`, functions properly. The line it was replaced with, `*r = *r.WithContext()`, changes the request that all downstream middleware have pointers to. This is a potential race risk, and provides them access to values that they do not need (after all, it was an upstream middleware that added the value) and, maybe, should not see at all.